internal/util/basesubjectset.go

```diff
@@ -224,32 +244,581 @@ func (bss BaseSubjectSet[T]) UnionWithSet(other BaseSubjectSet[T]) {
+		if existingExclusion.GetSubjectId() == concrete.GetSubjectId() {
+			// The conditional expression for the exclusion is a combination on the existing exclusion or
+			// the new expression.
+			exclusionConditionalExpression := combine(existingExclusion.GetConditionalExpression(), concrete.GetConditionalExpression())
+
+			updatedExclusions = append(updatedExclusions, wst.constructor(
+				concrete.GetSubjectId(),
+				exclusionConditionalExpression,
+				nil,
+				existingExclusion,
+				concrete),
+			)
+			wasFound = true
+		} else {
+			updatedExclusions = append(updatedExclusions, existingExclusion)
+		}
+	}
+
+	if !wasFound {
+		updatedExclusions = append(updatedExclusions, concrete)
+	}
+
+	constructed := wst.constructor(
+		tuple.PublicWildcard,
+		existing.GetConditionalExpression(),
+		updatedExclusions,
+		existing)
+	wst.wildcard = &constructed
+}
+
+func (wst *wildcardSubjectTracker[T]) intersectDiffWithWildcard(other *wildcardSubjectTracker[T]) {
+	if wst.wildcard == nil {
+		// Nothing to do if no wildcard present.
+		return
+	}
+
+	// If the other wildcard does not exist, change this wildcard to empty.
+	if other.wildcard == nil {
+		wst.wildcard = nil
+		return
+	}
+
+	// If the other wildcard exists, then the intersection between the two wildcards is an && of
+	// their conditionals, and a *union* of their exclusions.
+	existing := *wst.wildcard
+	foundWildcard := *other.wildcard
+
+	concreteExclusions := newConcreteSubjectSet[T](wst.constructor)
+	for _, excludedSubject := range existing.GetExcludedSubjects() {
+		concreteExclusions.unionConcrete(excludedSubject)
+	}
+
+	for _, excludedSubject := range foundWildcard.GetExcludedSubjects() {
+		concreteExclusions.unionConcrete(excludedSubject)
+	}
+
+	constructed := wst.constructor(
+		tuple.PublicWildcard,
+		and(existing.GetConditionalExpression(), foundWildcard.GetConditionalExpression()),
+		concreteExclusions.asSlice(),
+		existing,
+		foundWildcard)
+	wst.wildcard = &constructed
+}
+
+// exclusionsMap returns a map from subject ID to the associated exclusion, for each exclusion
+// on this wildcard.
+func (wst *wildcardSubjectTracker[T]) exclusionsMap() map[string]T {
+	if wst.wildcard == nil {
+		return nil
+	}
+
+	return exclusionsMapFor(*wst.wildcard)
+}
+
+// exclusionsMapFor creates a map of all the exclusions on a wildcard, by subject ID.
+func exclusionsMapFor[T Subject[T]](wildcard T) map[string]T {
+	exclusions := make(map[string]T, len(wildcard.GetExcludedSubjects()))
+	for _, excludedSubject := range wildcard.GetExcludedSubjects() {
+		exclusions[excludedSubject.GetSubjectId()] = excludedSubject
+	}
+	return exclusions
+}
+
+// combine combines two caveat expressions into one. If one of the expressions is empty, then
+// the entire expression is *short circuited*, and a nil is returned. Otherwise, the expressions
+// are union'ed together.
+func combine(first *v1.CaveatExpression, second *v1.CaveatExpression) *v1.CaveatExpression {
```

Comment:
a little confusing that this is called by `union*` methods, when there's a `union` with slightly different behavior below.

Reply:
Because in the case where we have a subject without a conditional expression at all, we need to treat the entire expression as short circuited. Its not a union: its a short circuiting union

Reply:
I understand what `combine` is doing differently from `union`, I'm just pointing out naming ambiguity

Reply:
Oh, agreed. Any ideas for a better/good name?

Reply:
Renamed